yjit/src/mm.rs

```diff
@@ -0,0 +1,376 @@
+pub struct VirtualMemory<A: Allocator> {
+    /// Location of the virtual memory region.
+    region_start: *mut u8,
+
+    /// Size of the region in bytes.
+    region_size_bytes: usize,
+
+    /// Number of bytes per "page", memory protection permission can only be controlled at this
+    /// granularity.
+    page_size_bytes: usize,
+
+    /// Number of bytes that have we have allocated physical memory for starting at
+    /// [Self::region_start].
+    mapped_region_bytes: usize,
+
+    /// Keep track of the address of the last written to page.
+    /// Used for changing protection to implement W^X.
+    current_write_page: Option<usize>,
+
+    /// Zero size member for making syscalls to get physical memory during normal operation.
+    /// When testing this owns some memory.
+    allocator: A,
+}
+
+/// Groups together the two syscalls to get get new physical memory and to change
+/// memory protection. See [VirtualMemory] for details.
+pub trait Allocator {
+    #[must_use]
+    fn mark_writable(&mut self, ptr: *const u8, size: u32) -> bool;
+
+    fn mark_executable(&mut self, ptr: *const u8, size: u32);
+}
+
+/// Pointer into a [VirtualMemory].
+/// We may later change this to wrap an u32.
+/// Note: there is no NULL constant for CodePtr. You should use Option<CodePtr> instead.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Debug)]
+#[repr(C)]
+pub struct CodePtr(*const u8);
+
+/// Errors that can happen when writing to [VirtualMemory]
+#[derive(Debug, PartialEq)]
+pub enum WriteError {
+    OutOfBounds,
+    FailedPageMapping,
+}
+
+use WriteError::*;
+
+impl<A: Allocator> VirtualMemory<A> {
+    /// Bring a part of the address space under management.
+    pub fn new(allocator: A, page_size: u32, virt_region_start: *mut u8, size_bytes: usize) -> Self {
+        assert_ne!(0, page_size);
+        let page_size_bytes = page_size.as_usize();
+
+        Self {
+            region_start: virt_region_start,
+            region_size_bytes: size_bytes,
+            page_size_bytes,
+            mapped_region_bytes: 0,
+            current_write_page: None,
+            allocator,
+        }
+    }
+
+    /// Return the start of the region as a raw pointer. Note that it could be a dangling
+    /// pointer so be careful dereferencing it.
+    pub fn start_ptr(&self) -> CodePtr {
+        CodePtr(self.region_start)
+    }
+
+    /// Size of the region in bytes where writes could be attempted.
+    pub fn virtual_region_size(&self) -> usize {
+        self.region_size_bytes
+    }
+
+    /// Write a single byte. The first write to a page makes it readable.
+    pub fn write_byte(&mut self, write_ptr: CodePtr, byte: u8) -> Result<(), WriteError> {
+        let page_size = self.page_size_bytes;
+        let raw: *mut u8 = write_ptr.raw_ptr() as *mut u8;
+        let page_addr = (raw as usize / page_size) * page_size;
+
+        if self.current_write_page == Some(page_addr) {
+            // Writing within the last written to page, nothing to do
```

Comment:
I'm a bit unclear as to how the page switching logic works.

Does the current_page_write change if you try to mark a page as executable and this triggers a protection fault or something?

Reply:
When the struct gets a request to write, it changes protection on that page and remember which page that is. Future requests to write to the same page don't need more syscalls. When we mark all the code executable, we do set `current_write_page` to `None`, so first request to write after making everything executable issues a syscall.

This is an extension on the old code's logic for handling W^X. The old code used `ALIGNED_WRITE_POSITION_NONE = 1` instead of `None` as a sentinel.

Reply:
But do you trap writes to unallocated space through protection errors?

Reply:
Right. Never written to regions aren't marked as executable or writable so they maintain their "no permission" protection.